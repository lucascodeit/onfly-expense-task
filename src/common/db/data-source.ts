import { DataSource } from 'typeorm'
import { join } from 'path'
import dotenv from 'dotenv'
import environment from '../environment/environment'

dotenv.config()

const parentDir = join(__dirname, '..', '..')

let AppDataSource: DataSource

export const dataSource = new DataSource({
  type: 'mysql',
  host: environment.DB_HOST,
  port: Number(environment.DB_PORT),
  username: environment.DB_USERNAME,
  password: environment.DB_PASSWORD,
  database: environment.DB_NAME,
  synchronize: false,
  logging: false,
  ssl: false,
  entities: [`${parentDir}/**/*-entity.{js,ts}`],
  migrations: [`${parentDir}/common/db/migrations/*.{js,ts}`]
})

export const getDataSoure = async () => {
  if (!AppDataSource) {
    AppDataSource = await dataSource.initialize()
  }
  return AppDataSource
}
