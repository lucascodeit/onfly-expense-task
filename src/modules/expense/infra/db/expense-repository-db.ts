import { injectable } from 'tsyringe'
import { Pagination } from '../../../../common/interfaces/pagination'
import { Expense } from '../../domain/expense'
import { ExpenseFilter } from '../../domain/expense-filter'
import { ExpenseRegister } from '../../domain/expense-register'
import { ExpenseRepository } from '../../domain/expense-repository'
import { getDataSoure } from '../../../../common/db/data-source'
import { ExpenseEntity } from './entities/expense-entity'
import { Repository } from 'typeorm'
import { ExpenseRegisterToExpenseEntity } from './adpater/expense-register-to-expense-entity'
import { ExpenseEntityToExpense } from './adpater/expense-entity-to-expense'
import { ExpenseFilterToWhere } from './adpater/expense-filter-to-where'

@injectable()
export class ExpenseRepositoryDb implements ExpenseRepository {
  async create(expense: ExpenseRegister): Promise<Expense> {
    const source = await this.getRepository()
    const expenseSaved = await source.save(ExpenseRegisterToExpenseEntity.of(expense))
    return ExpenseEntityToExpense.of(expenseSaved)
  }

  async update(filter: ExpenseFilter, data: ExpenseRegister): Promise<Expense> {
    const source = await this.getRepository()
    const expense = await this.getOne(filter)
    if (!expense) return null
    const dataForUpdate = ExpenseRegisterToExpenseEntity.of(data)
    const result = await source.update(expense.id, dataForUpdate)
    if (!result || result.affected < 1) {
      return null
    }
    return new Expense({
      id: expense.id,
      date: dataForUpdate.date,
      value: dataForUpdate.value,
      description: dataForUpdate.description,
      user: expense.user
    })
  }

  async delete(filter: ExpenseFilter): Promise<number> {
    const source = await this.getRepository()
    const expenseToUpdate = await this.getOne(filter)
    if (!expenseToUpdate) return null
    const result = await source.delete(expenseToUpdate.id)
    if (!result || result.affected < 1) {
      return null
    }
    return result.affected
  }

  async getList(filter: ExpenseFilter): Promise<Expense[]> {
    const source = await this.getRepository()
    const where = ExpenseFilterToWhere.of(filter).where
    const result = await source.find({ where, relations: ['user'] })
    if (!result) return null
    return result.map(ExpenseEntityToExpense.of)
  }

  async getOne(filter: ExpenseFilter): Promise<Expense> {
    const source = await this.getRepository()
    const where = ExpenseFilterToWhere.of(filter).where
    const result = await source.findOne({ where, relations: ['user'] })
    if (!result) return null
    return ExpenseEntityToExpense.of(result)
  }

  private async getRepository(): Promise<Repository<ExpenseEntity>> {
    const source = await getDataSoure()
    return source.getRepository(ExpenseEntity)
  }
}
