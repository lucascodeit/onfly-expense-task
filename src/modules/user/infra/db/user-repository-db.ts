import { Repository } from 'typeorm'
import { injectable } from 'tsyringe'
import { getDataSoure } from '../../../../common/db/data-source'
import { UserRepository } from '../../domain/user-repository'
import { UserRegister } from '../../domain/user-register'
import { User } from '../../domain/user'
import { UserEntity } from './entities/user-entity'
import { UserRegisterToUserEntity } from './adapter/user-register-to-user-entity'
import { UserEntityToUser } from './adapter/user-entity-to-user'

@injectable()
export class UserRepositoryDb implements UserRepository {
  async createUser(userRegister: UserRegister): Promise<{ id: number }> {
    const source = await this.getUserRepository()
    const userEntitiy = UserRegisterToUserEntity.of(userRegister)
    const userCreated = await source.save(userEntitiy)
    return { id: userCreated.id }
  }

  async getById(id: number): Promise<User> {
    const source = await this.getUserRepository()
    const user = await source.findOne({ where: { id } })
    if (user) {
      return UserEntityToUser.of(user)
    }
    return null
  }

  async getUserByEmail(email: string): Promise<User | null> {
    const source = await this.getUserRepository()
    const user = await source.findOne({ where: { email } })
    if (user) {
      return UserEntityToUser.of(user)
    }
    return null
  }

  async getUserRepository(): Promise<Repository<UserEntity>> {
    const source = await getDataSoure()
    return source.getRepository(UserEntity)
  }
}
