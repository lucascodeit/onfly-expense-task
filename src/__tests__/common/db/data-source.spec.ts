const mockMethod = jest.fn().mockImplementation(data => {})

jest.mock('typeorm', () => ({
  DataSource: class DataSourceMock {
    constructor(protected datasource: any) {}
    get data() {
      return this.datasource
    }
  }
}))
import { dataSource } from '../../../common/db/data-source'

describe('testing data-source from db', () => {
  it('should have the correct configurations', () => {
    expect((dataSource as any).data).toMatchSnapshot()
  })
})
