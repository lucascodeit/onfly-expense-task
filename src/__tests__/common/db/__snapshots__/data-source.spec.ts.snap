// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing data-source from db should have the correct configurations 1`] = `
{
  "database": "onfly_db",
  "entities": [
    "/Users/bagyadmin/Projects/onfly/src/**/*-entity.{js,ts}",
  ],
  "host": "localhost",
  "logging": false,
  "migrations": [
    "/Users/bagyadmin/Projects/onfly/src/common/db/migrations/*.{js,ts}",
  ],
  "password": "mysql-onfly",
  "port": 3306,
  "ssl": false,
  "synchronize": false,
  "type": "mysql",
  "username": "root",
}
`;
